clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java

```diff
@@ -0,0 +1,146 @@
+ */
+
+package org.jboss.as.clustering.controller;
+
+import java.util.EnumSet;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import javax.xml.stream.XMLStreamException;
+
+import org.jboss.as.controller.Extension;
+import org.jboss.as.controller.ExtensionContext;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.PersistentResourceXMLDescription;
+import org.jboss.as.controller.parsing.ExtensionParsingContext;
+import org.jboss.as.controller.persistence.SubsystemMarshallingContext;
+import org.jboss.dmr.ModelNode;
+import org.jboss.staxmapper.XMLElementReader;
+import org.jboss.staxmapper.XMLElementWriter;
+import org.jboss.staxmapper.XMLExtendedStreamReader;
+import org.jboss.staxmapper.XMLExtendedStreamWriter;
+
+/**
+ * Generic extension implementation that registers a single subsystem.
+ * @author Paul Ferraro
+ */
+public class SubsystemExtension<S extends Enum<S> & Schema<S>> implements Extension {
+
+    private final String name;
+    private final Model currentModel;
+    private final Supplier<ManagementRegistrar<SubsystemRegistration>> registrarFactory;
+    private final S currentSchema;
+    private final Function<S, XMLElementReader<List<ModelNode>>> readerFactory;
+    private final XMLElementWriter<SubsystemMarshallingContext> writer;
+    private final XMLElementReader<List<ModelNode>> currentReader;
+
+    /**
+     * Constructs a new extension using a reader/writer factory.
+     * @param <P> the parser type
+     * @param name the subsystem name
+     * @param currentModel the current model
+     * @param registrarFactory a factory for creating the subsystem resource registrar
+     * @param currentSchema the current schema
+     * @param descriptionFactory an XML description factory
+     */
+    public SubsystemExtension(String name, Model currentModel, Supplier<ManagementRegistrar<SubsystemRegistration>> registrarFactory, S currentSchema, Function<S, PersistentResourceXMLDescription> descriptionFactory) {
+        // Build xml description for current schema only once
+        this(name, currentModel, registrarFactory, currentSchema, descriptionFactory, descriptionFactory.apply(currentSchema));
+    }
+
+    private SubsystemExtension(String name, Model currentModel, Supplier<ManagementRegistrar<SubsystemRegistration>> registrarFactory, S currentSchema, Function<S, PersistentResourceXMLDescription> descriptionFactory, PersistentResourceXMLDescription currentDescription) {
+        // Reuse current xml description between reader and writer
+        this(name, currentModel, registrarFactory, currentSchema, descriptionFactory.andThen(XMLDescriptionReader::new), new XMLDescriptionWriter(currentDescription), new XMLDescriptionReader(currentDescription));
+    }
+
+    /**
+     * Constructs a new extension using a reader factory and a separate writer implementation.
+     * @param <P> the parser type
```

Comment:
```suggestion
     * @param <S> the parser type
```

Though, unfortunately Javadoc doesn't allow for that at all here.

```
Generating /tmp/org/jboss/as/clustering/controller/SubsystemExtension.html...
/Users/rhusar/git/wildfly/clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java:60: error: @param name not found
     * @param <S> the parser type
               ^
```